Remove redundant `#available(iOS 13, *)` checks

mokagio opened the pull request:
They always return `true` since the deployment version is 13.0, and will soon become 14.0.

Only one has been left, in `WidgetStyles.swift`, because it will change again when bumping the deployment target to 14+.

I was looking at #available statements as part of https://github.com/wordpress-mobile/WordPress-iOS/pull/19496 and realized we still has some of these laying around.

See also #19511.

## Regression Notes

1. Potential unintended areas of impact – N.A.
2. What I did to test those areas of impact (or what existing automated tests I relied on) – N.A.
3. What automated tests I added (or what prevented me from doing so) – N.A.

---

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes. **N.A.**
- [x] I have considered adding accessibility improvements for my changes. **N.A.**
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary. **N.A.**

mokagio committed:
Remove redundant `#available(iOS 13, *)` checks

They always return `true` since the deployment version is 13.0, and will
soon become 14.0.

Only one has been left, in `WidgetStyles.swift`, because it will change
again when bumping the deployment target to 14+.

AliSoftware commented on WordPress/Classes/ViewRelated/Blog/Blog + Me/GravatarButtonView.swift:
```diff
@@ -66,13 +66,7 @@ extension GravatarButtonView {
-
-            return .gray(.shade20)
-        }
+        static var color: UIColor = .separator
```
At that point I think it can be a `static let`?
```suggestion
        static let color: UIColor = .separator
```

AliSoftware commented on WordPress/Classes/ViewRelated/Gutenberg/Views/GutenGhostView.swift:
```diff
@@ -93,12 +93,7 @@ class GutenGhostView: UIView {
-        }
-        return .clear
-    }
+    static var ghostToolbarBackground: UIColor = UIColor(light: .clear, dark: UIColor.colorFromHex("2e2e2e"))
```
```suggestion
    static let ghostToolbarBackground: UIColor = UIColor(light: .clear, dark: UIColor.colorFromHex("2e2e2e"))
```

PS: I'm surprised that you can do that in an `extension` as you can't add instance variable / storage in `extension`… but not sure about add `static` ones on the type, so maybe that's ok because of `static`? (I must be rusty in Swift already…)

If that's _not_ allowed, I'd assume that restriction applies equally to `static let` and `static var` storages, and so neither code would compile, and the solution would be to go back to a computed variable `static var ghostToolbarBackground: UIColor { .init(…) }` instead. But CI seems to be compiling the code happily, so, that's probably ok after all?

mokagio replied:
Yes, it is allowed.

I wouldn't be surprised if it was a recent Swift change though, because I remember running into the constraint of not being able to add stored properties in `extension` in the past. _Or, maybe I never tried to do it with `static`._

mokagio replied:
> I wouldn't be surprised if it was a recent Swift change though, because I remember running into the constraint of not being able to add stored properties in extension in the past. _Or, maybe I never tried to do it with `static`._

Might actually be a case of me never running into this with `static` or always assuming it worked the same as with instance. There is [at least one StackOverflow answer](https://stackoverflow.com/a/33603246/809944) using `static left` in an extension, dated 2014.

mokagio replied:
Oh, and by the way, stored properties are still not allowed:

<img width="1132" alt="image" src="https://user-images.githubusercontent.com/1218433/198162821-bd595c1c-83d1-42a3-a773-10cdb68ec1ef.png">

AliSoftware replied:
Ah, yeah, that makes sense I guess.

I totally get why there is this limitation for `var` instance variables. But given that `let` are constants, and thus not taking storage space on the instance and thus not modifying the underlying memory layout of an instance of the type, I wonder why instance constants `let` are not allowed tbh (after all, they could be seen as synonyms to `var potato: String { "potato" }`, at least in behavior…).

[Maybe this will be allowed in a future version of Swift though](https://forums.swift.org/t/allow-stored-properties-in-extensions/60889)?

AliSoftware commented on WordPress/Classes/ViewRelated/Reader/Select Interests/ReaderInterestsStyleGuide.swift:
```diff
@@ -36,13 +36,7 @@ class ReaderInterestsStyleGuide {
-
-        return .white
-    }
+    public static var buttonContainerViewBackgroundColor: UIColor = .tertiarySystemBackground
```
```suggestion
    public static let buttonContainerViewBackgroundColor: UIColor = .tertiarySystemBackground
```

AliSoftware commented on WordPress/Classes/ViewRelated/Stats/Extensions/WPStyleGuide+Stats.swift:
```diff
@@ -262,12 +262,7 @@ extension WPStyleGuide {
-            }
-            return .neutral(.shade10)
-        }
+        static var mapBackground: UIColor = .systemGray4
```
```suggestion
        static let mapBackground: UIColor = .systemGray4
```

AliSoftware commented on WordPress/Classes/ViewRelated/Stats/Today Widgets/WidgetStyles.swift:
```diff
@@ -8,13 +8,7 @@ class WidgetStyles: NSObject {
-            return .divider
-        }
-    }()
+    static var separatorColor: UIColor = .separator
```
```suggestion
    static let separatorColor: UIColor = .separator
```

mokagio committed:
Convert a few `static var` to `static let`

Co-authored-by: Olivier Halligon <olivier.halligon@automattic.com>

mokagio committed:
Update all computed `static` colors to `static let` in `GutenGhostView`

Maybe due to a recent Swift improvement, we are now able to declare
`static var` and `static let` within an `extension`. There's no need to
compute the value at runtime anymore.

AliSoftware commented on WordPress/WordPressShareExtension/MainShareViewController.swift:
```diff
@@ -96,14 +96,8 @@ private extension MainShareViewController {
-            shareNavController.transitioningDelegate = extensionTransitioningManager
-            shareNavController.modalPresentationStyle = .custom
-        }
+        // We need to make sure we don't end up with stacked modal view controllers by using this:
+        shareNavController.modalPresentationStyle = .overFullScreen
```
This is not an equivalent conversion, as you lost the `if editorController.originatingExtension == .saveToDraft` condition.

mokagio replied:
Great catch! Thanks!

mokagio replied:
@AliSoftware I addressed this in [`6d49db7` (#19513)](https://github.com/wordpress-mobile/WordPress-iOS/pull/19513/commits/6d49db787d1d52b2ed0643b02a21e66e030eb73e), thanks again.

 I started looking at how to add a unit test to catch
this issue but I run into a number of problems.

1. The code is in the WordPressShareExtension target and we cannot test
   that one directly; Xcode doesn't let you create test target against
   app extensions
2. Even if we could access the code, the logic depends on
   `editorController`, an instance property that sets the state based
   whether `Bundle.main.bundleIdentifier` contains "DraftAction". That
   is something we cannot modify at test run time.
3. Plus a few access level constraints, but those could be easily changed. I don't think we need to worry much about `private` vs `internal` in the context of an app extension.

In short, to test this trivial behavior, we'll need to put in place a
far-from trivial test harness. The idea I have at the moment would be to
extract as much of the logic shared between WordPressShareExtension,
WordPressDraftActionExtension, and their two Jetpack counter parts into
a framework, which we could then test in isolation and design in such a
way that we don't need to inspect the bundle identifier at runtime.

It's worth saying that folks were aware of the code limitations at the
time of implementing the behavior but must have been pressed for time to
ship a bug fix, as evidenced by the comment in the `editorController`
implementation.

mokagio committed:
Re-introduce accidentally removed check in `MainShareViewController`

This occurred while removing `#available(iOS 13, *)` checks a few
commits before this. I was too eager in my deletions.

See discussion at
https://github.com/wordpress-mobile/WordPress-iOS/pull/19513#discussion_r1006679109

For the record, I started looking at how to add a unit test to catch
this issue but I run into a number of problems.

1. The code is in the WordPressShareExtension target and we cannot test
   that one directly; Xcode doesn't let you create test target against
   app extensions
2. Even if we could access the code, the logic depends on
   `editorController`, an instance property that sets the state based
   whether `Bundle.main.bundleIdentifier` contains "DraftAction". That
   is something we cannot modify at test run time.

In short, to test this trivial behavior, we'll need to put in place a
far-from trivial test harness. The idea I have at the moment would be to
extract as much of the logic shared between WordPressShareExtension,
WordPressDraftActionExtension, and their two Jetpack counter parts into
a framework, which we could then test in isolation and design in such a
way that we don't need to inspect the bundle identifier at runtime.

It's worth saying that folks were aware of the code limitations at the
time of implementing the behavior but must have been pressed for time to
ship a bug fix, as evidenced by the comment in the `editorController`
implementation.

AliSoftware commented on WordPress/WordPressShareExtension/MainShareViewController.swift:
```diff
@@ -96,13 +96,9 @@ private extension MainShareViewController {
+        if editorController.originatingExtension == .saveToDraft {
+            // We need to make sure we don't end up with stacked modal view controllers by using this:
             shareNavController.modalPresentationStyle = .overFullScreen
-        } else {
-            shareNavController.transitioningDelegate = extensionTransitioningManager
-            shareNavController.modalPresentationStyle = .custom
```
@mokagio  you still need to keep this `else` clause and code path for the case when the source is not the SaveToDraft extension

mokagio replied:
_See why I need tests?!?!_

AliSoftware replied:
Oh wait, and now I just read the `// code comment` that was there before your change, and that said:
```
// iOS 13 has proper animations and presentations for share and action sheets. So the `else` case should be removed when iOS 13 is minimum.
```

So maybe you were right to remove the `else` clause after all! 🤔 (so sorry for the back-and-forth 😅  I should be more mindful of code comments in diffs)

Though that comment still begs the question, if we remove that `else` clause, are we also supposed to remove the `if editorController.originatingExtension == .saveToDraft` as well, and use the same `modalPresentationStyle = .custom` for all cases? Or should we still keep the `if` condition and only use `.custom` for SaveToDraft but not to other cases?

@bjtitus, as the author of that code back in https://github.com/wordpress-mobile/WordPress-iOS/pull/13146 (even if this was 3 years ago and you moved to other products since…), maybe you remember?

mokagio replied:
The release notes in that PR list

> * Fixed bugs with the "Save as Draft" action extension's navigation bar colors and iPad sizing in iOS 13.

Skimming the PR description, I think the `editorController.originatingExtension == .saveToDraft` check is intentional and required.

I'll verify the behavior for the two extensions and in both versions of the code and report. _Stay tuned._

mokagio replied:
With the code at [`cd07203` (#19513)](https://github.com/wordpress-mobile/WordPress-iOS/pull/19513/commits/cd0720341688986aaabba72be007583cb607d9eb), I'm getting the following

**Save as Draft**

<table>
<tr>
	<td>

![IMG_4733](https://user-images.githubusercontent.com/1218433/205038233-73872a9d-0130-4749-9342-c061d09e050c.PNG)

</td>
	<td>

![IMG_4734](https://user-images.githubusercontent.com/1218433/205038246-26d7b735-0942-40b7-b1b6-efb93d3b17d5.PNG)

</td>
</tr>
</table>

**Share**

<img alt="IMG_4736" src="https://user-images.githubusercontent.com/1218433/205038441-d31bfe38-1425-4a65-97bb-49778c0c84a8.PNG" width=300/>

Share doesn't look right. Might need more work. The first thing I'll try will be removing the `else` branch and add some break points. I'm also not sure why I'm getting that error, given the app was setup alright.

mokagio replied:
> Share doesn't look right. Might need more work.

https://github.com/wordpress-mobile/WordPress-iOS/pull/19700

mokagio committed:
Restore accidentally removed `else` branch in `MainShareViewController`

See discussion at
https://github.com/wordpress-mobile/WordPress-iOS/pull/19513#discussion_r1007829599

mokagio commented:
Closing in favor of #19699 and #19700.
